airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/cursor.py

```diff
@@ -0,0 +1,153 @@
+
+
+class Comparable(Protocol):
+    """Protocol for annotating comparable types."""
+
+    @abstractmethod
+    def __lt__(self: "Comparable", other: "Comparable") -> bool:
+        pass
+
+
+class CursorField:
+    def __init__(self, path: List[str]) -> None:
+        self._path = path
+
+    def extract_value(self, record: Record) -> Comparable:
+        return _extract_value(record.data, self._path)  # type: ignore  # we assume that the value the path points at is a comparable
+
+
+class Cursor(ABC):
+    @abstractmethod
+    def observe(self, record: Record) -> None:
+        """
+        Indicate to the cursor that the record has been emitted
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def close_partition(self, partition: Partition) -> None:
+        """
+        Indicate to the cursor that the partition has been successfully processed
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def end_sync(self) -> None:
+        """
+        Indicate to the cursor that the sync has been successfully processed
+        """
+        raise NotImplementedError()
+
+
+class NoopCursor(Cursor):
+    def observe(self, record: Record) -> None:
+        pass
+
+    def close_partition(self, partition: Partition) -> None:
+        pass
+
+    def end_sync(self) -> None:
+        pass
+
+
+class ConcurrentCursor(Cursor):
+    _START_BOUNDARY = 0
+    _END_BOUNDARY = 1
+
+    def __init__(
+        self,
+        stream_name: str,
+        stream_namespace: Optional[str],
+        stream_state: Any,
+        message_repository: MessageRepository,
+        connector_state_manager: ConnectorStateManager,
+        cursor_field: CursorField,
+        slice_boundary_fields: Optional[Tuple[str, str]],
+    ) -> None:
+        self._stream_name = stream_name
+        self._stream_namespace = stream_namespace
+        self._message_repository = message_repository
+        self._connector_state_manager = connector_state_manager
+        self._cursor_field = cursor_field
+        self._slice_boundary_fields = slice_boundary_fields if slice_boundary_fields else tuple()
+        self._most_recent_record: Optional[Record] = None
+
+        # TODO to migrate state. The migration should probably be outside of this class. Impact of not having this:
+        #  * Given a sync that emits no records, the emitted state message will be empty
+        self._state = {  # type: ignore  # waiting for when we implement the state migration
+            "slices": []  # empty for now but should look like `{start: 1, end: 10, parent_id: "id1"}`
+        }
+
+    def observe(self, record: Record) -> None:
+        if self._slice_boundary_fields:
+            # Given that slicing is done using the cursor field, we don't need to observe the record as we assume slices will describe what
+            # has been emitted. Assuming there is a chance that records might not be yet populated for the most recent slice, use a lookback
```

Comment:
This is an interesting point which I haven't thought about. My first reasoning is that this is part of the connector logic as the lookback window is part of the slicing logic which is [in the legacy stream](https://github.com/airbytehq/airbyte/blob/978a37e6598bace49f8ca23dcc3c3d04e2aeaeaa/airbyte-integrations/connectors/source-stripe/source_stripe/streams.py#L258) right now. Could this be part of the state logic? I don't see why not and this would have the benefit of being generic for all streams/sources

Reply:
+1 to adding to the shared logic (fine to leave as a follow up). Today we re-implement the same lookback window logic in multiple connectors which feels silly

Reply:
Issue created [here](https://github.com/airbytehq/airbyte/issues/31963)